sprints/dashboard/utils.py

```diff
@@ -97,6 +101,67 @@ def get_cell_member_names(conn: CustomJira, members: Iterable[str]) -> Dict[str,
     """Returns cell members with their names."""
     return {conn.user(member).displayName: member for member in members}
 
+def get_cell_member_roles(raise_exception: bool = False) -> Dict[str, List[str]]:
+    """Return a dictionary of cell members and their associated roles.
+    Example return: {'John Doe': ['Recruitment Manager', 'Sprint Planning Manager',...],...}.
+
+    Note: We have room for error here if member names are spelled differently, hence we must
+    be careful that the handbook & Jira have consistent naming.
+    """
+
+    # Example HTML: <li><a href="../roles/#cell-manager-recruitment">Recruitment manager</a>: John Doe</li>
+    ROLES_REGEX = r"<li>.*roles.*>([A-Za-z ]+).*: (.+)<\/li>"
+
+    r = requests.get(HANDBOOK_ROLES_PAGE)
+
+    # roles = [('Recruitment manager', 'John Doe'),('Sprint Planning Manager', 'John Doe'),...]
+    roles = re.findall(ROLES_REGEX, r.text)
+
+    # roles_dict = {'John Doe': ['Recruitment Manager', 'Sprint Planning Manager',...],...}
+    roles_dict = defaultdict(list)
+    for role, member in roles:
+        roles_dict[member].append(role)
+
+    # It's difficult to know whether we have read all roles
+    # So we assume that if we've read roles for less than 12 members, something's probably wrong
+    if raise_exception and len(roles_dict) == 0:
+        raise NoRolesFoundException("No roles were found at the handbook page: %s" % HANDBOOK_ROLES_PAGE)
+
+    return roles_dict
+
+def compile_participants_roles(
+    members: List[User],
+    rotations: Dict[str, List[str]],
+    cell_member_roles: Dict[str, List[str]]
+) -> Dict[str, List[str]]:
+    """Compile the final roles Dictionary from cell_member_roles and rotations data
+    """
+
+    roles: DefaultDict[str, List[str]] = defaultdict(list)
+    for member in members:
+        roles[member.emailAddress].extend(
+            cell_member_roles[member.displayName]
+        )
+
+        roles[member.emailAddress].extend(
+            get_rotations_roles_for_member(member.displayName, rotations)
+        )
+
+    return roles
+
+def get_rotations_roles_for_member(member_name: str, rotations: Dict[str, List[str]]):
+    """Given a member and a dictionary containing get_rotations_users() output, return a list
+    of all roles for that user.
+    """
+
+    roles = []
+    for duty, assignees in rotations.items():
+        for idx, assignee in enumerate(assignees):
+            # The rotations sheet sometimes contains only the first name
```

Comment:
Typo: `sometimes contains` -> `might contain`.

Reply:
39ab8d5

Reply:
Note: this is not the case anymore, as we've updated the page yesterday and now it contains full names.